src/Compilers/Core/Portable/Serialization/SerializationThreadPool.cs

```diff
@@ -0,0 +1,170 @@
+
+            /// <summary>The queue of work items. Also used as a lock to protect all relevant state.</summary>
+            private static readonly Queue<(Delegate, object?, TaskCompletionSource<object?>)> s_queue = new();
+
+            /// <summary>The number of threads currently waiting for work to arrive.</summary>
+            private static int s_availableThreads = 0;
+
+            /// <summary>
+            /// Queues an <see cref="Action"/> delegate to be executed immediately on another thread,
+            /// and returns a <see cref="Task"/> that represents its eventual completion. The task will
+            /// always end in the <see cref="TaskStatus.RanToCompletion"/> state; if the delegate throws
+            /// an exception, it'll be allowed to propagate on the thread, crashing the process.
+            /// </summary>
+            public static Task<object?> QueueAsync(Func<object?> threadStart) => QueueAsync((Delegate)threadStart, null);
+
+            /// <summary>
+            /// Queues an <see cref="Action{Object}"/> delegate and associated state to be executed immediately on another thread,
+            /// and returns a <see cref="Task"/> that represents its eventual completion. The task will
+            /// always end in the <see cref="TaskStatus.RanToCompletion"/> state; if the delegate throws
+            /// an exception, it'll be allowed to propagate on the thread, crashing the process.
+            /// </summary>
+            public static Task<object?> QueueAsync(Func<object?, object?> threadStart, object? state) => QueueAsync((Delegate)threadStart, state);
+
+            private static Task<object?> QueueAsync(Delegate threadStart, object? state)
+            {
+                // Create the TaskCompletionSource used to represent this work.
+                // Call sites only care about completion, not about the distinction between
+                // success and failure and do not expect exceptions to be propagated in this manner,
+                // so only SetResult is used.
+                var tcs = new TaskCompletionSource<object?>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+                // Queue the work for a thread to pick up. If no thread is immediately available, it will create one.
+                enqueue((threadStart, state, tcs));
+
+                // Return the task.
+                return tcs.Task;
+
+                static void createThread()
+                {
+                    // Create a new background thread to run the work.
+                    var t = new Thread(() =>
+                    {
+                        // Repeatedly get the next item and invoke it, setting its TCS when we're done.
+                        // This will wait for up to the idle time before giving up and exiting.
+                        while (tryDequeue(out (Delegate action, object? state, TaskCompletionSource<object?> tcs) item))
```

Comment:
```suggestion
                        while (tryDequeue(out var item))
```

Reply:
The 'Use var' analyzer and Quick Info both think these are not the same type.

Reply:
Not sure. I replaced it manually and it seems to work, but when I hover over `var` it shows all the types as fully-qualified.

Reply:
I wonder if we have an issue here with local functions... Strange